File: debuggingExerciseStarters/tipCalculator.py
import math

def calculateTip(billTotal, tipPercent, roundup=False):
    '''
Compute the tip based on the provided amount and percentage.  If the roundUp flag is set,
the tip should round the total up to the next full dollar amount.
   billTotal - total of the purchase
   percentage - the percentage tip (e.g., .2 = 20%, 1 = 100%)
   roundUp - if true, add to the tip so the billTotal + tip is an even dollar amount
 returns the calculated tip
    '''
    tip = round(billTotal * tipPercent, 2)
   # print(tip)
    
    total = billTotal + tip

    if roundup == True:
        extraTip = math.ceil(total)
       # print(extraTip)
        tip1 = round(extraTip - total, 2)
        print(tip)
        print(tip1)
        tip = tip + tip1

    


    
   # tip = billTotal * tipPercent
   # tip = round(tip, 2)
   # print(tip)
   # if roundup == True:
  #      tip = round(tip, 2)
  #  else:
     #   pass
    return tip


#==================================================================================================
# Testing code below
def testBasicTip():
    results = ""
    TESTS = [(20, 0.15, 3.0), (100, 0.001, 0.1), (10, 1, 10)]
    for test in TESTS:
        actual = calculateTip(test[0], test[1])
        expectedHigh = test[2] + 0.01
        expectedLow  = test[2] - 0.01
        if actual <= expectedLow or actual >= expectedHigh:
            results += "Basic Test Failed: for a total of " + str(test[0]) + " and percentage " + str(test[1]) + ", I expected " + str(test[2]) + " but your code returned " + str(actual) + "\n"
    return results

def testRoundupTip():
    results = ""
    TESTS = [(10.14, 0.15, 1.86), (100.01, 0.001, 0.99), (20.01, 0.15, 3.99)]
    for test in TESTS:
        actual = calculateTip(test[0], test[1], True)
        expectedHigh = test[2] + 0.01
        expectedLow  = test[2] - 0.01
        if actual <= expectedLow or actual >= expectedHigh:
            results += "Roundup Test Failed: for a total of " + str(test[0]) + " and percentage " + str(test[1]) + ", I expected " + str(test[2]) + " but your code returned " + str(actual) + "\n"
    return results

result = ""
result += testBasicTip()
result += testRoundupTip()

if len(result) == 0:
    print("All Tests Passed!")
else:
    print(result)

File: JavaBasics/src/tips/TipCalculator.java
package tips;

import java.util.Arrays;

/**
 * @author <Jaime Castaneda>
 */
public class TipCalculator 
{
	/**
	 * Compute the tip based on the provided amount and percentage.  If the roundUp flag is set,
	 * the tip should round the total up to the next full dollar amount.
	 * @param amount - total of the purchase
	 * @param percentage - the percentage tip (e.g., .2 = 20%, 1 = 100%)
	 * @param roundUp - if true, add to the tip so the amount + tip is an even dollar amount
	 * @return
	 */
    public double computeTip(double billTotal, double percentage, boolean roundUp)
    {
    	
    	double tipAmount = 0;
    	
  
    	tipAmount = billTotal * percentage;
    	
    	
    	
    	return tipAmount;
    	
        
        
    }

    /**
     * Run a command line version of the tip calculator.  It does NOT prompt the user.  The user must call using 
     * the following arguments:
     * <REQUIRED>
     *   -price xx.xx => the -price flag followed by a price 
     *   -percent xx.x => the -percent flag followed by a percent tip (e.g., 15.5 would be a 15.5% tip)
     * <OPTIONAL>
     *   roundup => the flag indicating to add an extra amount to the tip to round up to the next dollar value
     * For example, 
     *   java tips.TipCalculator -price 10.00 -percent 25 
     * or
     *   java tips.TipCalculator -price 10.00 -percent 25 roundup
     * @param args
     */
    public static void not_main(String[] args) // This is so when you hit the run button it runs your test, it is just main in real Java programs
    {
        boolean roundUp = false;
        double price    = 10;
        double percent  = 25;
        
        System.out.println(Arrays.toString(args));
        price = Double.parseDouble(args[1]);
        percent = Double.parseDouble(args[3])/100;

        // Fill the values above from the command line parameters (args) so the rest of this
        // call works.
         
        TipCalculator tc = new TipCalculator();
        double tip = tc.computeTip(price, percent, roundUp);
        double total = price + tip;
        System.out.printf("For $%.2f at a tip of %.1f%%%s\nTip: $%.2f\nTotal: $%.2f\n", 
                          price, percent*100, (roundUp? " rounded up": ""), tip, total);
    }
}